File: sdks/python/src/verdocs/utils/primitives.py
"""Name and sequence primitives (js-sdk: Utils/Primitives.ts)."""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from .strings import capitalize


def _read(source: Any, field: str) -> Any:
    # The js-sdk takes structural objects; the Python equivalents are dicts
    # (raw payloads) and pydantic models (parsed ones), so we accept both.
    if isinstance(source, Mapping):
        return source.get(field)
    return getattr(source, field, None)


def integer_sequence(start: int, count: int) -> list[int]:
    """Return a list of consecutive integers, e.g. [start, start + 1, ...].

    Frequently useful in rendering when there is no source list to iterate.

    Args:
        start: The first value.
        count: How many values to produce.

    Returns:
        A list of `count` integers beginning at `start`.
    """
    return list(range(start, start + count))


def format_full_name(source: Any = None) -> str:
    """Format a profile-like record's full name, capitalized.

    Args:
        source: Anything carrying first_name and last_name, as keys or
            attributes (a dict, a Profile, a Recipient). None and missing
            names are treated as empty.

    Returns:
        "First Last", trimmed, so a single known name comes back alone and
        no names at all come back as "".
    """
    first = capitalize(_read(source, "first_name") or "")
    last = capitalize(_read(source, "last_name") or "")
    return f"{first} {last}".strip()


def format_initials(profile: Any = None) -> str:
    """Format a profile's initials, e.g. "T U", with "--" for no profile.

    Args:
        profile: Anything carrying first_name and last_name, as keys or
            attributes. The js-sdk assumes both names are present; we treat
            a missing one as empty rather than raising.

    Returns:
        The space-separated uppercase initials, or "--" when there is no
        profile to read.
    """
    if not profile:
        return "--"
    first = capitalize(_read(profile, "first_name") or "")
    last = capitalize(_read(profile, "last_name") or "")
    return f"{first[:1]} {last[:1]}"


def full_name_to_initials(name: str) -> str:
    """Generate suggested initials for a full name, e.g. "John Doe" yields "JD".

    Splits on single spaces and keeps each word's first character unchanged,
    matching the js-sdk (which does not uppercase here).

    Args:
        name: The full name.

    Returns:
        The concatenated first characters.
    """
    return "".join(word[:1] for word in name.split(" "))

File: sdks/csharp/src/Verdocs.Sdk/Utils/Primitives.cs
using System.Text;
using Verdocs.Models;

namespace Verdocs.Utils;

/// <summary>Name and sequence primitives (js-sdk: Utils/Primitives.ts).</summary>
public static class Primitives
{
    /// <summary>
    /// Returns a list of consecutive integers, e.g. [start, start + 1, ...]. Frequently
    /// useful in rendering when there is no source list to iterate.
    /// </summary>
    /// <param name="start">The first value.</param>
    /// <param name="count">How many values to produce.</param>
    /// <returns>A list of <paramref name="count"/> integers beginning at <paramref name="start"/>.</returns>
    /// <sdkOperation>primitive.integerSequence</sdkOperation>
    /// <sdkGroup>Primitive</sdkGroup>
    /// <sdkPage>Helpers</sdkPage>
    public static IReadOnlyList<int> IntegerSequence(int start, int count)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(count);

        var values = new int[count];
        for (var i = 0; i < count; i++)
        {
            values[i] = start + i;
        }

        return values;
    }

    /// <summary>
    /// Formats a profile-like record's full name, capitalized. The js-sdk takes a structural
    /// source object; pass the record's name fields directly, e.g.
    /// FormatFullName(profile.FirstName, profile.LastName).
    /// </summary>
    /// <param name="firstName">The first name; null is treated as empty.</param>
    /// <param name="lastName">The last name; null is treated as empty.</param>
    /// <returns>
    /// "First Last", trimmed, so a single known name comes back alone and no names at all
    /// come back as "".
    /// </returns>
    /// <sdkOperation>primitive.formatFullName</sdkOperation>
    /// <sdkGroup>Primitive</sdkGroup>
    /// <sdkPage>Helpers</sdkPage>
    public static string FormatFullName(string? firstName, string? lastName)
    {
        return (Strings.Capitalize(firstName ?? string.Empty) + " " + Strings.Capitalize(lastName ?? string.Empty)).Trim();
    }

    /// <summary>
    /// Formats a profile's initials, e.g. "T U", with "--" for no profile.
    /// </summary>
    /// <param name="profile">
    /// The profile to read. The js-sdk assumes both names are present; a missing one is
    /// treated as empty here rather than throwing.
    /// </param>
    /// <returns>The space-separated uppercase initials, or "--" when there is no profile to read.</returns>
    /// <sdkOperation>primitive.formatInitials</sdkOperation>
    /// <sdkGroup>Primitive</sdkGroup>
    /// <sdkPage>Helpers</sdkPage>
    public static string FormatInitials(Profile? profile)
    {
        if (profile is null)
        {
            return "--";
        }

        var first = Strings.Capitalize(profile.FirstName ?? string.Empty);
        var last = Strings.Capitalize(profile.LastName ?? string.Empty);
        return (first.Length > 0 ? first[..1] : string.Empty) + " " + (last.Length > 0 ? last[..1] : string.Empty);
    }

    /// <summary>
    /// Generates suggested initials for a full name, e.g. "John Doe" yields "JD". Splits on
    /// single spaces and keeps each word's first character unchanged, matching the js-sdk
    /// (which does not uppercase here).
    /// </summary>
    /// <param name="name">The full name.</param>
    /// <returns>The concatenated first characters.</returns>
    /// <sdkOperation>primitive.fullNameToInitials</sdkOperation>
    /// <sdkGroup>Primitive</sdkGroup>
    /// <sdkPage>Helpers</sdkPage>
    public static string FullNameToInitials(string name)
    {
        ArgumentNullException.ThrowIfNull(name);

        // A repeated space yields an empty word, which contributes nothing (in js, undefined
        // joins as "").
        var initials = new StringBuilder();
        foreach (var word in name.Split(' '))
        {
            if (word.Length > 0)
            {
                initials.Append(word[0]);
            }
        }

        return initials.ToString();
    }
}
